Features/task 2

kubmir opened the pull request.

Commit:
KA-227 Add Models assembly with INotesRepository and Note model

Commit:
KA-227 Refactor application to use Models assembly

Commit:
KA-227 Add NotesRepository initial file

Commit:
KA-227 Rename Notes.Models to Notes.Contracts

Commit:
KA-227 Refactor using due to change of Note's namespace

Commit:
KA-227 Implement Repository logic

Commit:
Merge branch 'features/task-1' into features/task-2

Commit:
KA-227 Fix test's using after merge

Commit:
KA-227 Install Unity

Commit:
KA-227 Add Unity dependency injection

Commit:
KA-227 Remove Unity.Mvc

Commit:
KA-227 Refactor tests to use NSubstitute

Commit:
KA-227 Fix namespace of Notes.Api in asax and App_Start

Commit:
KA-227 Refactor NotesControllerTest init

Commit:
KA-227 Refactor using of Guid in tests

Commit:
Merge branch 'features/task-1' into features/task-2

Commit:
KA-227 Update PUT method to be up to date with task-1 PUT method

Commit:
KA-227 Use NotesRouteName constant in Post method in NotesController

Commit:
Merge branch 'develop' into features/task-2

Commit:
KA-227 Fix issue with building - system.net.http binding

Commit:
KA-227 Change configuration of DI

Remove files UnityConfig and UnityWebApiActivator. Add UnityResolver and
UnityConfig which configures using of Unity container.

Commit:
KA-227 Add IBootstrapper

Commit:
KA-227 Implement Bootstrapper for DI

Commit:
KA-227 Add interface for url location helpers

Commit:
KA-227 Implement UrlLocationHelper and use it in controller

Commit:
KA-227 Update tests to use UrlLocationHelper

Commit:
KA-227 Add DI for UrlLocationHelper

Commit:
KA-227 Move UnityResolver to Api.Helpers

Commit:
KA-227 Fix namespace of UnityResolver

Commit:
KA-227 Simplify initialization of NotesControllerTests

Commit:
KA-227 Refactor UrlLocationHelper

Rename method GetUrl to GetUrlWithId, use DI for HttpRequestMessage in
UrlLocationHelper constructor, register type HttpRequesMessage in the
unity container

Commit:
KA-227 Add tests for UrlLocationHelper

Commit:
KA-227 Add interface for DI container

Commit:
KA-227 Add container for DI

Commit:
KA-227 Refactor application to use Unity only in Dependency assembly

Assemblies Api, Contracts, Repository do not contain reference to Unity
anymore, Unity is isolated in own assembly called Dependency. IUnityContainer
is replaced by IMyContainer in assemblies Api, Contracts and Repository.

Commit:
KA-227 Refactor UnityResolver to ContainerResolver

Commit:
KA-227 Fix build issues with wrong file included for compilation

Commit:
KA-227 Extract Api.Services in src and tests

Commit:
KA-227 Move NotesRepository to Repository folder

Commit:
KA-227 Remove unused references

Commit:
KA-227 Move UrlLocationHelperTest to Helpers folder

Commit:
KA-227 Move MyContainer to Containers folder

Commit:
KA-227 Fix build issues - packages

Commit:
KA-227 Rename (I)MyContainer to (I)DependencyContainer

Commit:
KA-227 Rename Helpers to Services

Commit:
KA-227 Move types registration to Api.Services

Commit:
KA-227 Rename folder ApiHelpers to ApiServices

vit-svoboda commented on NotesBackEnd/src/Notes.Api.Services/Dependency/ApiServicesTypesRegistration.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace Notes.Api.Services.Dependency
+{
+    public class ApiServicesTypesRegistration : IBootstrapper
```
Since you implement `IBootstrapper` here, I'd expect the class to be called `*Boostrapper`, most likely `ApiServicesBootstrapper`.

vit-svoboda commented on NotesBackEnd/src/Notes.Dependency/Containers/DependencyContainer.cs:
```diff
@@ -0,0 +1,70 @@
+
+        public IDependencyContainer RegisterType<TFrom>()
+        {
+            _unityContainer.RegisterType<TFrom>(new InjectionFactory(GetHttpRequestMessage));
```
Any type I try to register, every time I get HttpRequestMessage? That doesn't seem right.

kubmir replied:
My solution is not the best one but I did not find any other way to obtain not null HttpRequestMessage, I tried to place GetHttpRequestMessage method to DependencyInjectionConfig and also to ApiServicesBootstrapper but it returned null all the time.

So now, I have own method for registration of HttpRequestMessage in DependencyContainer :/

vit-svoboda replied:
Well, your `RegisterType<TType>` could accept a factory function as a parameter. Something like this:
```
public IDependencyContainer RegisterType<TType>(Func<TType> factory) {
    _unityContainer.RegisterType<TType>(new InjectionFactory(factory));

    return this;
}
```
Then you'd pass the `GetHttpRequestMessage` method inside in the  type registration call.

kubmir replied:
I tried this before my dirty solution, however, InjectionFactory requires method which takes IUnityContainer as a parameter. 

        public InjectionFactory(Func<IUnityContainer, object> factoryFunc);


In my solution Unity is separated in assembly Notes.Dependency but I call registration of HttpRequestMessage in another assembly - I am not sure if it is a good idea to include Unity in different assemblies or register HttpRequestMessage in Notes.Dependency.

vit-svoboda commented on NotesBackEnd/src/Notes.Dependency/Containers/DependencyContainer.cs:
```diff
@@ -0,0 +1,70 @@
+            _unityContainer = container;
+        }
+
+        public IDependencyContainer RegisterType<TFrom>()
```
Maybe the `TFrom` is just a `TType` or something. There's no `TTo` here.

vit-svoboda commented on NotesBackEnd/src/Notes.Api.Services/Services/ContainerResolver.cs:
```diff
@@ -0,0 +1,43 @@
+
+        public ContainerResolver(IDependencyContainer container)
+        {
+            Container = container ?? throw new ArgumentNullException(nameof(container));
```
Cool use of C# 6 and 7 👍

vit-svoboda commented on NotesBackEnd/src/Notes.Api/App_Start/DependencyInjectionConfig.cs:
```diff
@@ -0,0 +1,23 @@
+        }
+
+        private static IDependencyContainer RegisterDependency(this IDependencyContainer container, IBootstrapper registrationClass)
+            => registrationClass.RegisterType(container);
```
This kind of looks like you register `container` to `registrationClass`. Perhaps the name of `RegisterType` could be a bit better.

vit-svoboda commented on NotesBackEnd/src/Notes.Api/Controllers/NotesController.cs:
```diff
@@ -2,37 +2,59 @@
+        {
+            var notes = await _repository.GetAllNotesAsync();
+
+            return await Task.FromResult(Ok(notes));
```
No need to `await` here. By making the method `async` you automatically return `Task` with anything you actually return. You only need `await` at the repository call since it's another async method. Here, you can just `return Ok(notes);`

vit-svoboda commented on NotesBackEnd/src/Notes.Api/Notes.Api.csproj:
```diff
@@ -50,27 +50,32 @@
+    <Reference Include="Microsoft.Web.Infrastructure, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>..\..\packages\Microsoft.Web.Infrastructure.1.0.0.0\lib\net40\Microsoft.Web.Infrastructure.dll</HintPath>
+    </Reference>
+    <Reference Include="netstandard" />
```
I don't think you want to reference .NETStarndard - you could, but doesn't really make sense at the entry point (Notes.Api project). Targetting .NETStandar in your service/repository libraries would make more sense since it would allow you to reference them in both .NET framework and .NET core projects. You don't need that, though ;)

vit-svoboda commented on NotesBackEnd/src/Notes.Contracts/ApiServices/IUrlLocationHelper.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public interface IUrlLocationHelper
+    {
+        String GetUrlWithId(string routeName, Guid id);
```
The location helper is just a thin wrapper around the `UrlHelper` from WebApi. I think it could take the responsibility for route names, but then you'd need to extract the route name from the controller. So I'm not really sure how good an idea it is.

kubmir replied:
I am not sure if I fulfill criteria right now - I added routeDictionary which contains both route and routeName and controller get value from it.

vit-svoboda commented on NotesBackEnd/tests/Notes.Api.Tests/packages.config:
```diff
@@ -1,12 +1,8 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+  <package id="Castle.Core" version="4.2.0" targetFramework="net471" />
```
You need Castle? 😮

kubmir replied:
When I try to remove Castle.Core I get an error message - Unable to uninstall "Castle.Core.4.2.0" because NSibstitute.3.1.0 depends on it - so I Think I really need it :)

vit-svoboda reviewed (dismissed):
Looks very nice, I think you're quite unsure with `async` `await`, though.

Commit:
KA-227 Rename *TypesRegistration to *Bootstrapper

Commit:
KA-227 Fix usage of Tasks in controller

Commit:
KA-227 Fix registration of types in DependencyContainer

Commit:
KA-227 Remove reference to netstandard in Api

Commit:
KA-227 Rename registrationClass to bootstrapper in DIConfig

Commit:
KA-227 Introduce RouteDictionary

Commit:
KA-227 Fix IDependencyContainer to extend IDisposable

Commit:
KA-227 Fix getting of HttpRequestMessage for container registration

Commit:
KA-227 Move bootstrapper's registration to DependencyContainerBuilder

Commit:
KA-227 Rewrite simple methods to use lambdas

Commit:
KA-227 Add lifetimeManager to DependencyContainer

Commit:
KA-227 Refactor usage of lifetime managers

Commit:
KA-227 Add separated interfaces for ContainerResolver and ContainerRegister

In this commit IDependencyContainer is divided into IDependencyContainerResolver
and IDependencyContainerRegister. This commit also refactor usage of RegisterType
methods and registration of HttpRequestMessage.

Commit:
KA-227 Refactor UrlLocationHelper with tests

Commit:
KA-227 Rename RouteDictionary to RoutesRegister

Commit:
KA-227 Clean solution - remove unneccessary packages

Commit:
KA-227 Refactor using of routes

Commit:
KA-227 Make NotesRepositary singletone per application

Commit:
KA-227 Clean solution

Commit:
KA-227 Fix generic issue in RegisterType<TTo>

Commit:
KA-227 Fix issue with missing id in url

Commit:
KA-227 Change lifetime managers in registration of types

Commit:
KA-227 Fix mocking of UrlHelper in UrlLocationHelper tests

Commit:
KA-227 Make RouteManager per application singleton

Commit:
KA-227 Remove Arg.Any from tests mocks

PetrSvirak commented on NotesBackEnd/src/Notes.Dependency/ContainersBuilders/DependencyContainerBuilder.cs:
```diff
@@ -0,0 +1,24 @@
+            return container;
+        }
+
+        private static IDependencyContainerRegister RegisterDependency(this IDependencyContainerRegister container, IBootstrapper bootstrapper)
```
Make it generic and save yourself the repeated initialization 😉

PetrSvirak commented on NotesBackEnd/src/Notes.Dependency/Containers/DependencyContainer.cs:
```diff
@@ -0,0 +1,64 @@
+        }
+
+        public IEnumerable<object> ResolveAll(Type serviceType)
+           => _unityContainer.ResolveAll(serviceType);
```
Have a look at the implementation. This can actually throw the very same `Exception`s as the `_unityContainer.Resolve`…

PetrSvirak commented on NotesBackEnd/src/Notes.Dependency/Containers/DependencyContainer.cs:
```diff
@@ -0,0 +1,64 @@
+            {
+                return _unityContainer.Resolve(serviceType);
+            }
+            catch (Exception)
```
This catch is too omnipotent (it masks too many exception/errors/problems).  Only justifiable ["catch 'em all"](http://wiki.c2.com/?PokemonExceptionHandling) is at the very top-level code that handles error page in IIS or main program of a console app with TUI… 

Here, you should catch only `ResolutionFailedException` and not even every `ResolutionFailedException` - only such that belongs to (system.web) contracts that are intentionally not fulfilled - ask around about this (me, Vašek, Matúš and Rado should all have this in fresh memory 😃)

PetrSvirak commented on NotesBackEnd/src/Notes.Api.Services/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,39 @@
+[assembly: AssemblyTrademark("")]
+[assembly: AssemblyCulture("")]
+
+[assembly: InternalsVisibleTo("Notes.Api")]
```
No, no, no, no 😉 
Only use `InternalsVisibleTo` for tests or legacy code you cannot refactor.

There is another way to provide the `Container` with the implementation of an `API` specific data class (or more precisely instance). Discuss this with Vašek or me, please.

PetrSvirak commented on NotesBackEnd/src/Notes.Api.Services/app.config:
```diff
@@ -0,0 +1,11 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
You should not be needing this file here. Only the `app.config` (or `web.config`) of an executable assembly is used by the framework (for example `tests` are executable as well as the web app)

PetrSvirak reviewed (changes requested):
I know you played around with the code for quite some time now, but wow. Just wow. It is really neat and well structured. I love it 😃 There are some pitfalls you did not dodge, but I'm sure they will all be fun to fix.

Commit:
KA-227 Refactor DependencyContainer generic types

Commit:
KA-227 Make RegisterDependency method generic

Commit:
KA-227 Remove app.config from Notes.Api.Services

Commit:
KA-227 Add tests for registered types in DependencyContainerBuilder

Commit:
KA-227 Remove internals visible to Notes.Api from Notes.Api.Services

Move RouteManager to Notes.Api assembly, refactor registration and resolve
of ContainerResolver

Commit:
KA-227 Refactor catching of exception during resolve

Commit:
KA-227 Create DependencyContainer interface and use it

PetrSvirak commented on NotesBackEnd/src/Notes.Dependency/ContainersBuilders/DependencyContainerBuilder.cs:
```diff
@@ -12,13 +12,16 @@ public static IDependencyContainerResolver SetUpContainer()
 
-        private static IDependencyContainerRegister RegisterDependency(this IDependencyContainerRegister container, IBootstrapper bootstrapper)
-            => bootstrapper.RegisterType(container);
+        private static IDependencyContainerRegister RegisterDependency<T>(this IDependencyContainerRegister container) where T : IBootstrapper, new()
```
I don't like the simple "T" - it is variable like any other (kind of, for the naming purposes). Always give type variables purposeful names - here it could be `TBootrapper` it will make the implementation a bit more readable. It is also OK to call `=> new TBootrstrapper().RegisterType(container)` – see how well-named generic type makes the code great(ly readable) again? :)

PetrSvirak reviewed (approved):
Job well done 😃  There is but one detail in naming, please fix in one of the following tasks.
